### PR TITLE
Develop/kwargs

### PR DESCRIPTION
**Pull request by mbourqui**
Update download() signature to use keyword arguments instead of positional arguments. For example, the downloader can now be called like
```
import ncbi_genome_download as ngd
ngd.download(group='bacteria', assembly_level='complete')
```
Argument parser still works and uses same default values as method signature. Default values are stored in a customized Enum to ensure immutability and provide some ease of use.

Tests are also updated.

### CODE REVIEW

File: ncbi_genome_download/__init__.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
     EDefaults
 )
 
-__version__ = '0.2.3'
+__version__ = '0.2.4'
```
Comment (kblin) [inline ncbi_genome_download/__init__.py]: Please don't bump the version number in an unrelated commit.

File: ncbi_genome_download/core.py [hunk 2]
```diff
@@ -1,74 +1,193 @@
...
 import requests
 
+from ncbi_genome_download.summary import SummaryReader
+
 # Python < 2.7.9 hack: fix ssl support
 if sys.version_info < (2, 7, 9):  # pragma: no cover
     from requests.packages.urllib3.contrib import pyopenssl
-    pyopenssl.inject_into_urllib3()
 
+    pyopenssl.inject_into_urllib3()
```
Comment (kblin) [inline ncbi_genome_download/core.py]: I'd prefer to leave the inject close to the import, as that keeps the "hack old pythons to properly support ssl" block.

File: ncbi_genome_download/core.py [hunk 3]
```diff
@@ -1,74 +1,193 @@
...
-}
-
-
-DownloadJob = namedtuple('DownloadJob', ['full_url', 'local_file', 'expected_checksum', 'symlink_path'])
-
-
-def download(args):
-    '''Download data from NCBI'''
+@unique
+class EMap(Enum):
```
Comment (kblin) [inline ncbi_genome_download/core.py]: Nice use of an Enum here. I'm not used to a bunch of the new Python 3 features yet, but this does make the code easier to read.

File: ncbi_genome_download/core.py [hunk 4]
```diff
@@ -174,7 +190,11 @@ def download(**kwargs):
     taxid = kwargs.pop('taxid', EDefaults.TAXID.default)
     human_readable = kwargs.pop('human_readable', False)
     parallel = kwargs.pop('parallel', EDefaults.PROCESSES.default)
-    # TODO: warning/error if unrecognized option
+    # FIXME: improve error handling and feedback
+    assert len(kwargs) == 0, "Unrecognized option(s)"
+    if len(kwargs) > 0:
```
Comment (kblin) [inline ncbi_genome_download/core.py]: That code is unreachable, as the line above already exits if `len(kwargs) != 0`.

File: tests/test_core.py [hunk 5]
```diff
@@ -19,175 +20,176 @@ def req():
...
-    fake_args = Namespace(section='refseq', domain='all', uri=core.NCBI_URI,
-                          output='/tmp/fake', file_format='genbank', assembly_level='all',
-                          genus='', species_taxid=None, taxid=None, human_readable=False,
-                          parallel=1)
-    core.download(fake_args)
-    assert _download_mock.call_count == len(core.SUPPORTED_DOMAINS)
+    download_mock = mocker.MagicMock()
+    monkeypatch.setattr(core, 'download', download_mock)
+    core.download(domain='all', output='/tmp/fake')
+    assert download_mock.call_count == len(core.SUPPORTED_DOMAINS)
```
Comment (kblin) [inline tests/test_core.py]: Does this one still pass? It seems that in the current code, the `download` function is only called once.

File: tests/test_core.py [hunk 6]
```diff
@@ -369,22 +378,28 @@ def test_grab_checksums_file(req):
...
 42c1bb1447aea2512a17aeb3645b55e9  ./GCF_000009605.1_ASM960v1_genomic.fna.gz
 8a685d49d826c4f0ad05152e906f3250  ./GCF_000009605.1_ASM960v1_genomic.gbff.gz
 e2d9e1cfa085cb462a73d3d2d2c22be5  ./GCF_000009605.1_ASM960v1_genomic.gff.gz
 d8ce7c80d457e012f9d368a4673dea65  ./GCF_000009605.1_ASM960v1_protein.faa.gz
 This_is_totally_an_invalid_line!
 620a09de4286f66113317456c0dc8f66  ./GCF_000009605.1_ASM960v1_protein.gpff.gz
-'''
+"""
     expected = [
-        {'checksum': 'd3c2634cedd0efe05cbf8a5f5384d921', 'file': 'GCF_000009605.1_ASM960v1_feature_table.txt.gz'},
```
Comment (kblin) [inline tests/test_core.py]: I know editors don't like long lines, but in this case I feel this change hurts the readability, rather than helping.

File: ncbi_genome_download/__main__.py [hunk 7]
```diff
@@ -9,21 +9,21 @@
 def main():
     """Build and parse command line"""
     parser = argparse.ArgumentParser()
-    parser.add_argument('domain',
-                        choices=dflt.DOMAIN.choices,
-                        default=dflt.DOMAIN.default,
-                        help='The NCBI "domain" to download (default: %(default)s)')
+    parser.add_argument('group',
```
Comment (kblin) [inline ncbi_genome_download/__main__.py]: NCBI calls this "domain", so I'd like to stick to that term, even if it's wrong.

**Comment by mbourqui** [comment]
The first time I tried the tests locally, I had some path issues. So I moved the test folder in 2e2f7e3f. But I moved it back to its initial location in 49d6cd8 as the tests were then running fine again.

File: tests/test_core.py (hunk 6 above)
Comment (mbourqui) [inline tests/test_core.py]: Was done automatically by my IDE. Fixed it.

File: ncbi_genome_download/core.py (hunk 2 above)
Comment (mbourqui) [inline ncbi_genome_download/core.py]: Done by the auto code reformatting feature of my IDE. Fixed it.

File: ncbi_genome_download/core.py (hunk 3 above)
Comment (mbourqui) [inline ncbi_genome_download/core.py]: Thank you!

File: ncbi_genome_download/core.py (hunk 4 above)
Comment (mbourqui) [inline ncbi_genome_download/core.py]: Dropped unreachable code.

File: tests/test_core.py (hunk 5 above)
Comment (mbourqui) [inline tests/test_core.py]: According to http://github.drone.secondarymetabolites.org/kblin/ncbi-genome-download/44, this test passes.
`download` is called once, but `_download` is called once for each taxonomic group.

Comment (mbourqui) [inline tests/test_core.py]: Indeed, in that code version, the test was not passing. It is now passing again, after being updated to the changes of `core.py`.

File: ncbi_genome_download/__main__.py (hunk 7 above)
Comment (mbourqui) [inline ncbi_genome_download/__main__.py]: I found in that document ftp://ftp.ncbi.nlm.nih.gov/pub/factsheets/HowTo_Downloading_Genomic_Data.pdf (in the *Genomes FTP Directory Structure* section) that they were refering to it as taxonomic group, thus my naming change.
But I can revert to domain if you'd like to.

File: ncbi_genome_download/__init__.py (hunk 1 above)
Comment (mbourqui) [inline ncbi_genome_download/__init__.py]: Fixed in 5fb4e40.

File: ncbi_genome_download/__main__.py (hunk 7 above)
Comment (kblin) [inline ncbi_genome_download/__main__.py]: I can't find the README file I had for this. I guess the NCBI gets to rename their stuff if they like. For sure "taxonomic group" is a more correct term anyway. Let's just keep it for now.

**Comment by kblin** [comment]
Ok, playing with this a bit more, it seems that now the command line tool doesn't work anymore. I should have a fix for that, though.

**Comment by mbourqui** [comment]
Try with my latest commit

File: ncbi_genome_download/__main__.py [hunk 8]
```diff
@@ -2,9 +2,9 @@
 import argparse
 import logging
 
-from . import __version__
-from . import download
-from . import EDefaults as dflt
+from ncbi_genome_download import __version__
+from ncbi_genome_download import download
+from ncbi_genome_download import EDefaults as dflt
```
Comment (kblin) [inline ncbi_genome_download/__main__.py]: Where are you seeing this as a problem? That bit wasn't the issue for me.
Rather `download(vars(args))` clearly needs to be `download(**vars(args))`, and then you also need to remove the "debug", "verbose", and "retries" entry.

Comment (mbourqui) [inline ncbi_genome_download/__main__.py]: I encountered locally the issue I fixed. Maybe it just affects Python 3.
Working on the other issues.

Comment (kblin) [inline ncbi_genome_download/__main__.py]: Possibly. I started testing on 2.7 :)

**Comment by mbourqui** [comment]
Should be fine now!

**Comment by kblin** [comment]
It also looks like the default file format changed from genbank to cds-fasta.

**Comment by kblin** [comment]
That's on python 2.7 only, though.

**Comment by kblin** [comment]
Yeah, cPython does not guarantee that the order of dictionaries is preserved. It's an implementation detail that this happens in cPython 3.6.

**Comment by mbourqui** [comment]
Ouch, that's an issue. I was using Enums because of the ordering garantee...

**Comment by kblin** [comment]
Well, looking at the enum34 code, it seems to be using an OrderedDict, so I assume there might be an issue in `Emap.keys()`

File: ncbi_genome_download/core.py [hunk 9]
```diff
@@ -97,6 +97,8 @@ def get_content(cls, key):
 
 
 class EFormats(EMap):
+    __order__ = 'GENBANK FASTA FEATURES GFF PROTFASTA GENREPT WGS CDSFASTA RNAFASTA'
+    # only needed in Python 2
```
Comment (kblin) [inline ncbi_genome_download/core.py]: Maybe put the comment above the line if it doesn't fit at the end?

**Comment by kblin** [comment]
Yup, this works, thanks a bunch.

File: ncbi_genome_download/core.py (hunk 9 above)
Comment (mbourqui) [inline ncbi_genome_download/core.py]: Certainly!

**Comment by kblin** [comment]
Thanks for the contribution, and your fast response.

**Comment by mbourqui** [comment]
You're welcome, thank you for your base code! :)
